src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -19,73 +19,55 @@
     {
-        private readonly (string Source, string Name)[] _columns;
-        private readonly IHost _host;
-
         public CopyColumnsEstimator(IHostEnvironment env, string input, string output) :
             this(env, (input, output))
         {
         }
 
         public CopyColumnsEstimator(IHostEnvironment env, params (string source, string name)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(CopyColumnsEstimator)), new CopyColumnsTransform(env, columns))
         {
-            Contracts.CheckValue(env, nameof(env));
-            _host = env.Register(nameof(CopyColumnsEstimator));
-            _host.CheckValue(columns, nameof(columns));
-            var newNames = new HashSet<string>();
-            foreach (var column in columns)
-            {
-                if (!newNames.Add(column.name))
-                    throw Contracts.ExceptUserArg(nameof(columns), $"New column {column.name} specified multiple times");
-            }
-            _columns = columns;
         }
 
-        public CopyColumnsTransform Fit(IDataView input)
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
         {
-            // Invoke schema validation.
-            GetOutputSchema(SchemaShape.Create(input.Schema));
-            return new CopyColumnsTransform(_host, _columns);
-        }
+            Host.CheckValue(inputSchema, nameof(inputSchema));
 
-        public SchemaShape GetOutputSchema(SchemaShape inputSchema)
-        {
-            _host.CheckValue(inputSchema, nameof(inputSchema));
             var resultDic = inputSchema.Columns.ToDictionary(x => x.Name);
-            foreach (var column in _columns)
+            foreach (var column in Transformer.Columns)
             {
                 if (!inputSchema.TryFindColumn(column.Source, out var originalColumn))
-                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", column.Source);
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", column.Source);
                 var col = new SchemaShape.Column(column.Name, originalColumn.Kind, originalColumn.ItemType, originalColumn.IsKey, originalColumn.Metadata);
                 resultDic[column.Name] = col;
             }
             return new SchemaShape(resultDic.Values);
         }
     }
 
-    public sealed class CopyColumnsTransform : ITransformer, ICanSaveModel
+    public sealed class CopyColumnsTransform : OneToOneTransformerBase
     {
-        private readonly (string Source, string Name)[] _columns;
-        private readonly IHost _host;
         public const string LoaderSignature = "CopyTransform";
         private const string RegistrationName = "CopyColumns";
         public const string Summary = "Copy a source column to a new column.";
         public const string UserName = "Copy Columns Transform";
         public const string ShortName = "Copy";
```

Comment:
can the public members be made internal? if they are public, they will display in the documentation site.

Reply:
these are now internal

---
In reply to: [222532040](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222532040) [](ancestors = 222532040)